lldb/source/Target/Language.cpp

```diff
@@ -574,3 +576,164 @@ bool SourceLanguage::IsObjC() const {
 bool SourceLanguage::IsCPlusPlus() const {
   return name == llvm::dwarf::DW_LNAME_C_plus_plus;
 }
+
+// BEGIN SWIFT
+// Implement LanguageCPlusPlus::GetParentNameIfClosure and upstream this.
+// rdar://152321823
+
+/// If `sc` represents a "closure"-like function (according to
+/// Language::GetParentNameIfClosure), returns sc.function and all parent
+/// functions up to and including the first non-closure-like function. If `sc`
+/// is not a closure, or if the query does not make sense for `language`,
+/// returns a list containing only sc.function.
+static llvm::SmallVector<Function *>
+GetParentFunctionsWhileClosure(const SymbolContext &sc,
+                               const Language &language) {
+  // The algorithm below terminates on the assumption that
+  // `GetParentNameIfClosure` produces an empty string when composing that
+  // function with itself enough times. For safety, define an upper limit.
+  constexpr auto upper_limit = 8;
+
+  llvm::SmallVector<Function *> parents;
+  Function *root = sc.function;
+  if (root == nullptr)
+    return parents;
+
+  parents.push_back(root);
+  for (int idx = 0; idx < upper_limit; idx++) {
+    ConstString mangled = root->GetMangled().GetMangledName();
+    std::string parent = language.GetParentNameIfClosure(mangled);
+    if (parent.empty())
+      break;
+
+    // Find the enclosing function, if it exists.
+    SymbolContextList sc_list;
+    Module::LookupInfo lookup_info(
+        ConstString(parent), lldb::FunctionNameType::eFunctionNameTypeFull,
+        lldb::eLanguageTypeSwift);
+    sc.module_sp->FindFunctions(lookup_info, CompilerDeclContext(),
+                                ModuleFunctionSearchOptions(), sc_list);
+    if (sc_list.GetSize() != 1 || sc_list[0].function == nullptr)
+      break;
+    parents.push_back(sc_list[0].function);
+    root = sc_list[0].function;
+  }
+  return parents;
+}
+
+/// Scans the line table of `function` looking for the first entry whose line
+/// number is `line_number`. If no such entry is found, returns the entry
+/// closest to but after `line_number`.
+static std::optional<Address> FindAddressForLineNumber(Function &function,
```

Comment:
When upstreaming, this might be a better fit as a member function of lldb::Function

Reply:
Yup, good point